File: phantom-type/haskell/authentication.hs
-- runghc ./authentication.hs
-- echo ./authentication.hs | entr -c runghc /

import Text.Printf (printf)

data Validated
data Unverified
data Reason = InsufficientLength
newtype Password validation = Password String

fromString :: String -> Password Unverified
fromString = Password

validate :: Password Unverified -> Either Reason (Password Validated)
validate (Password pw)
    | length pw > 1 = Right (Password pw)
    | otherwise = Left InsufficientLength

stringOfReason :: Reason -> String
stringOfReason InsufficientLength = "Not long enough"

main :: IO ()
main = do
    let pass = fromString "p@a$$w0rd"
    case validate pass of
        Right (Password str) -> printf "Password check OK: %s!\n" str
        Left err -> printf "Password rejected: %s\n" (stringOfReason err)

File: phantom-type/ocaml/lib/authentication.ml
type validated
type unverified
type reason = Insufficient_length
type 'validation password = Password of string

let from_string (str : string) : unverified password = Password str

let validate (Password pw : unverified password) : (validated password, reason) result =
  if String.length pw > 0 then
    Ok (Password pw)
  else
    Error Insufficient_length
;;

let string_of_reason = function
  | Insufficient_length -> "Not long enough"
;;
